src/pycapsule.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+use crate::Python;
+use crate::{ffi, AsPyPointer, PyAny};
+use crate::{pyobject_native_type_core, PyErr, PyResult};
+use std::ffi::{c_void, CStr};
+use std::os::raw::c_int;
+
+/// Represents a Python Capsule
+/// As described in [Capsules](https://docs.python.org/3/c-api/capsule.html#capsules)
+/// > This subtype of PyObject represents an opaque value, useful for C extension
+/// > modules who need to pass an opaque value (as a void* pointer) through Python
+/// > code to other C code. It is often used to make a C function pointer defined
+/// > in one module available to other modules, so the regular import mechanism can
+/// > be used to access C APIs defined in dynamically loaded modules.
+///
+///
+/// # Example
+/// ```
+///  use std::ffi::CString;
+///  use pyo3::{prelude::*, pycapsule::PyCapsule};
+///
+///  #[repr(C)]
+///  struct Foo {
+///     pub val: u32,
+///  }
+///
+///  let r = Python::with_gil(|py| -> PyResult<()> {
+///     let foo = Foo { val: 123 };
+///     let name = CString::new("builtins.capsule").unwrap();
+///
+///     let capsule = PyCapsule::new(py, foo, name.as_ref())?;
+///
+///     let module = PyModule::import(py, "builtins")?;
+///     module.add("capsule", capsule)?;
+///
+///     let cap: &Foo = unsafe { PyCapsule::import(py, name.as_ref(), false)? };
+///     assert_eq!(cap.val, 123);
+///     Ok(())
+///  });
+///  assert!(r.is_ok());
+/// ```
+#[repr(transparent)]
+pub struct PyCapsule(PyAny);
+
+pyobject_native_type_core!(PyCapsule, ffi::PyCapsule_Type, #checkfunction=ffi::PyCapsule_CheckExact);
+
+impl PyCapsule {
+    /// Constructs a new capsule of whose contents are `T` associated with `name`.
+    pub fn new<'py, T: 'static>(py: Python<'py>, value: T, name: &CStr) -> PyResult<&'py Self> {
```

Comment:
Doesn't this need at least a `T: Send` bound?  Once given to Python the object can be transferred to any thread, so both `import()` as well as the destructor can be called from any thread.

Reply:
Thanks for the review! 

I think this will handicap the functionality quite a bit. I personally, was hoping to use capsules to pass some functions, but [it doesn't appear fn is Send](https://doc.rust-lang.org/std/primitive.fn.html#)

Do you know how one would deal with this situation? (ie the current `test_pycapsule_func` then fails.

Reply:
I don't see why a function pointer wouldn't be `Send` (and the [playground seems to agree](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dbabdd6b7cebdd11ceedd4b9f2a2d3ae))...

I think the docs don't show the auto traits if they are derived. Only if you explicitly do an `impl !Send` or `impl Send where ...`.

Reply:
`fn` should be `Send` - and yes good catch, this should definitely require `Send`.

Reply:
Ah, I now see the original test was casting the pointer to `*const c_void`, which is not Send. Looking at e.g. https://internals.rust-lang.org/t/shouldnt-pointers-be-send-sync-or/8818 it seems that this is not really a question of safety.

The easiest workaround seems to be to make a struct `SendPtr(*const c_void)` with `#[repr(transparent)]` and have an `unsafe impl Send` for it. Maybe we can supply this in this module? Let's hear @davidhewitt 's opinion...

Reply:
`0bd03de` looks like it's solved this one by removing the cast?

Reply:
That one (just passing the function itself) gives a segfault now, which I haven't figured out yet.

Reply:
IMO you'll have to cast the function to `fn(u32) -> u32` when making the box. Function items (`foo` alone) are zero sized types, not function pointers, so the box will be "empty".

Since it never makes sense to put ZSTs into a capsule, we should at least document this and put an `assert!(size_of::<T>() > 0)` to panic at capsule creation if the invariant is violated.

Reply:
Great! I had no idea, casting it fixed it; thank you. Learning a lot here. :-) 
I've updated; for my own understanding, why is it preferred to panic instead of returning an error in this case?

Reply:
Just an instinct; this is a violation of the contract (at least when it's documented as such) and can never work, so failing fast would be my preference here.

It would of course be preferred to have a static assert that fires at compile time. There is no such facility in `std` yet, AFAIK.

Reply:
Well, it *is* possible: https://users.rust-lang.org/t/is-it-possible-to-assert-at-compile-time-that-foo-t-is-not-called-with-a-zst

Reply:
Cool! Added with https://github.com/PyO3/pyo3/pull/1980/commits/7a87a15760bedeeffb0fdaf47bc117a9f3548289, see what you think.